Pcap++/header/KniDevice.h

```diff
@@ -0,0 +1,709 @@
+			CALLBACK_MTU,
+			/** KniIoctlCallbacks#config_network_if and KniOldIoctlCallbacks#config_network_if callback */
+			CALLBACK_LINK,
+			/** KniIoctlCallbacks#config_mac_address callback */
+			CALLBACK_MAC,
+			/** KniIoctlCallbacks#config_promiscusity callback */
+			CALLBACK_PROMISC
+		};
+
+		/**
+		 * @brief New callbacks for KNI device events.
+		 * This structure MUST be used ONLY when KniDevice#callbackVersion returns
+		 * KniCallbackVersion#CALLBACKS_NEW.
+		 * Or if You are sure that DPDK version used is 17.11 or higher.
+		 * If some callback is not provided (NULL) the request will always succeeds
+		 * if other is not specified in callback description.
+		 */
+		struct KniIoctlCallbacks
+		{
+			/**
+			 * Pointer to function of changing MTU.
+			 * Must return 0 in case of success or negative error code
+			 */
+			int (*change_mtu)(uint16_t port_id, unsigned int new_mtu);
+			/**
+			 * Pointer to function of configuring network interface.
+			 * Must return 0 in case of success or negative error code
+			 */
+			int (*config_network_if)(uint16_t port_id, uint8_t if_up);
+			/**
+			 * Pointer to function of configuring mac address.
+			 * If callback is not provided and port_id of KNI device is not UINT16_MAX
+			 * then DPDK will use predefined callback witch sets MAC address of
+			 * DPDK port port_id via rte_eth_dev_default_mac_addr_set.
+			 * Must return 0 in case of success or negative error code
+			 */
+			int (*config_mac_address)(uint16_t port_id, uint8_t mac_addr[]);
+			/**
+			 * Pointer to function of configuring promiscuous mode.
+			 * If callback is not provided and port_id of KNI device is not UINT16_MAX
+			 * then DPDK will use predefined callback witch sets promiscuous mode of
+			 * DPDK port port_id via rte_eth_promiscuous_enable/rte_eth_promiscuous_disable.
+			 * Must return 0 in case of success or negative error code
+			 */
+			int (*config_promiscusity)(uint16_t port_id, uint8_t to_on);
+		};
+
+		/**
+		 * @brief Old callbacks for KNI device events.
+		 * This structure MUST be used ONLY when KniDevice#callbackVersion returns
+		 * KniCallbackVersion#CALLBACKS_OLD.
+		 * Or if You are sure that DPDK version used is lower than 17.11.
+		 * If some callback is not provided (NULL) the request will always succeeds.
+		 */
+		struct KniOldIoctlCallbacks
+		{
+			/**
+			 * Pointer to function of changing MTU.
+			 * Must return 0 in case of success or negative error code
+			 */
+			int (*change_mtu)(uint8_t port_id, unsigned int new_mtu);
+			/**
+			 * Pointer to function of configuring network interface.
+			 * Must return 0 in case of success or negative error code
+			 */
+			int (*config_network_if)(uint8_t port_id, uint8_t if_up);
+		};
+
+		/**
+		 * @brief KNI device initialization data.
+		 * Used to create new KNI device.
+		 * Usage of callbacks member or oldCallbacks member is defined by
+		 * result of KniDevice#callbackVersion
+		 */
+		struct KniDeviceConfiguration
+		{
+			enum
```

Comment:
If I have a choice i would use constexpr variables but we are in domain of C++98 where they are not existent.
The macro is an awful choice because it is like creating a global variable (except it much worse). Any code style guide will prohibit this kind of macro usage.
const variable is a subject of compiler optimisation and may not be optimized on low optimisation levels like -O0 and, ofcourse by the rules of C++ it mas have some address. Variables like such will accumulate over large code-base into a junk region in executable file that has no purpose.
The advantage of this approach is that on debug the symbol name for enum is saved for constants so it will be displayed even on old debugers and compilers. Other one is a strict naming rules (which is not for macro). End ofcourse enum value has no address and don't junks the executable.

Reply:
I do understand these reasons but still I think using an enum as a constant is not standard and many people will not understand it

Reply:
Ok maybe You will accept my point after this code sample:
https://godbolt.org/z/-ypdgE
Static const variable can be changed programmatically in runtime but emun value - only manualy thru code editing. I can put another example with inline assembly that changes this static variable to fuhrer proof my point but I see it as an overkill for our discussion.
Try to change optimization levels and see the results.
This is Undefined Behaviour in its best -> different compilers act differently.
But still this is NOT possible when the enum value is used.
As a concern about people understanding - low user knowledge about C++ behaviours is not a reason to write a bad code for more knowledgeable programmers. This technique is well known for the advantages I stated above.
So I prefer it to stay as the enum value.
Another reason to use it as the enum value is that static variable may or may not be used by compiler on constexpr evaluation where enum value is guaranteed to be used (the users of PcapPlusPlus is not only one that use C++98 but in all standarts and You should remember it).

Reply:
thanks @echo-Mike for the thorough explanation. I understand your concern about people being able to hack the static const variable and change its value. 

My only question is: why? why would someone want to hack it? this class is all about providing an easy way to set up KNI devices. I don't see any reason why someone would work hard and hack the system. And if they want to do it, they can do it in so many ways... they can also take the code (which is open source) and simply change the value...

My main objective in building this library is to provide APIs which are easy to understand and use. I'm less concerned about people trying to hack it

Reply:
Ok, let's take it easy. Maybe You grant to me an opportunity to teach the users of PcapPlusPlus with this C++ useful technic in following way.
I will create the doxygen note that describes this technic and the pros and cons of other one (like macro and const static) and in each place where it is used in header file I will give a doxygen link to that note in enum documentation.
This will provide users with comprehensive guide on some C++ behaviours and will teach them the good coding style.
In that way all other similar situations may be ported to this technic (like other static constants and macro) throughout all library.
Can You grant me a favour to do it?

Reply:
I guess I'm ok with that :)

Reply:
\-[To short on time to implement. Solution: moved to macro defs or deleted. Maybe some day I will return to this feature]